pkg/controller/elasticsearch/driver/upscale_state.go

```diff
@@ -5,62 +5,64 @@
 package driver
 
 import (
-	"sync"
+	"math"
 
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 
-	"github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1alpha1"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common"
-	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/nodespec"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 )
 
-type upscaleStateBuilder struct {
-	once         *sync.Once
-	upscaleState *upscaleState
-}
-
-func (o *upscaleStateBuilder) InitOnce(c k8s.Client, es v1alpha1.Elasticsearch, esState ESState) (*upscaleState, error) {
-	if o.once == nil {
-		o.once = &sync.Once{}
-	}
-	var err error
-	o.once.Do(func() {
-		o.upscaleState, err = newUpscaleState(c, es, esState)
-	})
-	return o.upscaleState, err
-}
-
 type upscaleState struct {
 	isBootstrapped      bool
 	allowMasterCreation bool
+	accountedNodes      int32
+	createsAllowed      int32
 }
 
-func newUpscaleState(c k8s.Client, es v1alpha1.Elasticsearch, esState ESState) (*upscaleState, error) {
+func newUpscaleState(
+	ctx upscaleCtx,
+	actualStatefulSets sset.StatefulSetList,
+	expectedResources nodespec.ResourcesList,
+) (*upscaleState, error) {
 	state := &upscaleState{
-		isBootstrapped:      AnnotatedForBootstrap(es),
+		isBootstrapped:      AnnotatedForBootstrap(ctx.es),
 		allowMasterCreation: true,
 	}
-	if !state.isBootstrapped {
-		return state, nil
-	}
-	// is there a master node creation in progress already?
-	masters, err := sset.GetActualMastersForCluster(c, es)
-	if err != nil {
-		return nil, err
-	}
-	for _, masterNodePod := range masters {
-		isJoining, err := isMasterNodeJoining(masterNodePod, esState)
+
+	if state.isBootstrapped {
+		// is there a master node creation in progress already?
+		masters, err := sset.GetActualMastersForCluster(ctx.k8sClient, ctx.es)
 		if err != nil {
 			return nil, err
 		}
-		if isJoining {
-			state.recordMasterNodeCreation()
+		for _, masterNodePod := range masters {
+			isJoining, err := isMasterNodeJoining(masterNodePod, ctx.esState)
+			if err != nil {
+				return nil, err
+			}
+			if isJoining {
+				state.recordMasterNodeCreation()
+			}
 		}
 	}
+
+	// calculate how many replicas can we create according to desired state and maxSurge
+	var createsAllowed int32 = math.MaxInt32
+	if ctx.es.Spec.UpdateStrategy.ChangeBudget != nil {
+		createsAllowed = int32(ctx.es.Spec.UpdateStrategy.ChangeBudget.MaxSurge)
+		createsAllowed += expectedResources.StatefulSets().ExpectedNodeCount()
+		createsAllowed -= actualStatefulSets.ExpectedNodeCount()
```

Comment:
I am a bit confused. Should this not be the `max(actualStatefulSets.ExpectedNodeCount(), expectedResources.StatefulSets().ExpectedNodeCount()) + maxSurge` How could we otherwise make progress, if we are going  to replace a stateful set?

Reply:
> 🤦‍♂ my bad and sorry, forgot to mention that there should be some data in the nodes

@barkbay no that's on me being a bit slow in the uptake :-) Am I right assuming that your suggestion of excluding nodes step by step as we remove them (which is of course correct) complemented by actually creating nodes at the same time. I am not sure if my interpretation of the surge budget of 1 is OK to say we can have one additional node in addition to what ever number of nodes exist currently in the cluster. 

Otherwise we would just needless migrate data to nodes that are about to be removed in the next round.

Reply:
> I am a bit confused. Should this not be the max(actualStatefulSets.ExpectedNodeCount(), expectedResources.StatefulSets().ExpectedNodeCount()) + maxSurge How could we otherwise make progress, if we are going to replace a stateful set?

I'd think that user applying a spec downscaling from 150 to 100 nodes and 20 surge expects to see at most 120 nodes instead of 170. User can make progress when surge is adjusted so it allows creating nodes despite actual state. Also, in any case, we can't be sure to make progress.

Reply:
I am willing to accept that there is no optimal solution that covers all cases. The question is how contrived the rename + scale down case is. Also I am bit worried that we don't surface the state of being stuck in this way  visible enough to the user. A log message might not be enough